WeakAuras/WeakAuras.lua

```diff
@@ -1642,9 +1642,127 @@ do
+      elseif theirs[key] ~= ourVal then
+        return true
+      end
+    end
+  end
+end
+
+local ignoredForDiffChecking = {
+  init_started = true,
+  yOffset = true,
+  xOffset = true,
+  anchorPoint = true,
+  id = true,
+  uid = true, -- honestly could leave it unignored but whatever
+  internalVersion = true,
+  width = true,
+  frameStrata = true,
+  anchorFrameType = true,
+  selfPoint = true,
+  height = true,
+  fixedWidth = true,
+  controlledChildren = true,
+}
+
+function WeakAuras.diff(ours, theirs, ignored)
+  if not ours or not theirs then return end
+  return recurseDiff(ours, theirs, ignoredForDiffChecking)
+end
+
+function WeakAuras.MatchData(data, children) 
+  local oldData, uids = db.displays[data.id], {}
+  if data.uid then uids[data.uid] = 0 end
+  local oldChildren
+  if children then
+    oldChildren = {}
+    for i, child in ipairs(children) do
+      if child.uid then uids[child.uid] = i end
+      oldChildren[i] = db.displays[child.id]
+    end
+  end
+  for id, existingData in pairs(db.displays) do
+    local uindex = existingData.uid and uids[existingData.uid]
+    if uindex == 0 then
+      match = true
+      oldData = existingData
+    elseif uindex then
+      match = true
+      oldChildren[uindex] = existingData
+    end
+  end
+  if match then
+    local modified = WeakAuras.diff(oldData, data)
+    if children then
+      local info = {
+        unmodified = oldData and not modified and 1 or 0,
+        modified = modified and 1 or 0,
+        added = not modified and not oldData and 1 or 0,
+        deleted = 0,
+      }
+      local oldTotal = oldData and oldData.controlledChildren and #oldData.controlledChildren or 0
+      for i, child in ipairs(children) do
+        if not oldChildren[i] then
+          modified = true
+          info.added = info.added + 1
+        else
+          if WeakAuras.diff(oldChildren[i], child) then
+            modified = true
+            info.modified = info.modified + 1
+            oldTotal = oldTotal - 1
+          else
+            child.same = true
+            oldTotal = oldTotal - 1
+            info.unmodified = info.unmodified + 1
+          end
+        end
+        info.deleted = math.max(oldTotal, 0)
+        modified = modified or info.deleted > 0
+      end
+      return oldData, oldChildren, (modified and info or nil)
+    else
+      return oldData, oldChildren, modified
+    end
+  end
+end
+
+local replaceFields
+function replaceFields(t1, t2, fields) 
+  -- this is vulnerable to a circular table, so
+  -- it is possible to trigger a stack overflow
+  for field, val in pairs(fields) do
+    if type(val) == "table" then
+      local tf1, tf2 = t1[field], t2[field]
+      if type(tf1) == "table" and type(tf2) == "table" then
+        replaceFields(tf1, tf2, val)
+      end
+    elseif val then
+      t1[field] = t2[field] or t1[field]
+    end
+  end
+end
+
+function WeakAuras.Update(oldData, newData, mode)
+  WeakAuras.DeleteOption(oldData)
```

Comment:
I'm not sure if that is the right thing to do if the id is unchanged.

Reply:
Frankly, aura modifying in general needs to be refactored. I can skip DeleteOption and instead Rename here, but we really need a less brutish method of modifying data.